File: offplan-escrow-mvp/datasources/warm.py
"""
Warm the public-data cache.

    DATA_OFFLINE=0 DATA_CONTACT="you@example.com" \
      python3 -m datasources.warm --site "Willow Park, Kilimani, Nairobi" \
                                  --lat -1.29210 --lon 36.78270

Fetches from Nominatim and Overpass, writes to fixtures/cache, and from then
on the app and the tests run offline against exactly what you fetched. Commit
the cache so a demo never depends on someone else's uptime.

    python3 -m datasources.warm --seed     write placeholder fixtures so the
                                           offline demo runs before you have
                                           fetched anything real
    python3 -m datasources.warm --stats    what is cached
"""

from __future__ import annotations

import argparse
import json
import os
import urllib.parse

from .http import CACHE_DIR, _cache_key, cache_stats
from .osm import NOMINATIM, OVERPASS


def warm(site: str, lat: float, lon: float, radius: int) -> None:
    from .osm import footprints, geocode

    print(f"geocoding  {site}")
    g = geocode(site)
    print(f"  {g.confidence:12} {g.latitude:.5f}, {g.longitude:.5f}  {g.display_name[:70]}")

    print(f"footprints within {radius}m of {lat}, {lon}")
    f = footprints(lat, lon, radius)
    print(f"  {f.building_count} buildings, {f.buildings_under_construction} under construction")
    if f.landuse:
        print(f"  land use: {', '.join(f.landuse)}")


def seed(site: str, lat: float, lon: float, radius: int) -> None:
    """
    Placeholder responses shaped exactly like the real APIs, so the offline
    demo runs end to end before anyone has network access. Overwritten the
    moment you warm the cache for real.
    """
    geo_url = NOMINATIM + "?" + urllib.parse.urlencode({
        "q": site, "countrycodes": "ke", "format": "jsonv2",
        "limit": 1, "addressdetails": 1,
    })
    geo_body = json.dumps([{
        "place_id": 297451102,
        "osm_type": "way",
        "osm_id": 1043318877,
        "lat": f"{lat + 0.0011:.7f}",
        "lon": f"{lon - 0.0008:.7f}",
        "display_name": f"{site}, Nairobi, Kenya",
        "category": "landuse",
        "type": "residential",
        "importance": 0.27,
        "address": {"suburb": "Kilimani", "city": "Nairobi", "country_code": "ke"},
    }]).encode()

    query = f"""
[out:json][timeout:40];
(
  way["building"](around:{radius},{lat},{lon});
  relation["building"](around:{radius},{lat},{lon});
  way["landuse"](around:{radius},{lat},{lon});
);
out center tags;
""".strip()
    over_body = json.dumps({
        "version": 0.6,
        "generator": "Overpass API",
        "elements": [
            {"type": "way", "id": 411229301,
             "center": {"lat": lat + 0.00042, "lon": lon + 0.00031},
             "tags": {"building": "apartments", "building:levels": "6"}},
            {"type": "way", "id": 411229774,
             "center": {"lat": lat - 0.00061, "lon": lon + 0.00090},
             "tags": {"building": "residential"}},
            {"type": "way", "id": 902114553,
             "center": {"lat": lat + 0.00105, "lon": lon - 0.00044},
             "tags": {"building": "construction", "construction": "apartments"}},
            {"type": "way", "id": 118844021,
             "tags": {"landuse": "residential", "name": "Kilimani"}},
        ],
    }).encode()

    for url, payload, body in [
        (geo_url, None, geo_body),
        (OVERPASS, urllib.parse.urlencode({"data": query}).encode(), over_body),
    ]:
        path = _cache_key(url, payload)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_bytes(body)
        print(f"seeded  {path.relative_to(CACHE_DIR.parent)}")


def main() -> None:
    p = argparse.ArgumentParser(description="Warm the public-data cache")
    p.add_argument("--site", default="Willow Park Block A, Kilimani")
    p.add_argument("--lat", type=float, default=-1.29210)
    p.add_argument("--lon", type=float, default=36.78270)
    p.add_argument("--radius", type=int, default=200)
    p.add_argument("--seed", action="store_true", help="write placeholder fixtures")
    p.add_argument("--stats", action="store_true", help="show cache contents")
    a = p.parse_args()

    if a.stats:
        s = cache_stats()
        print(f"{s['entries']} entries, {s['bytes']:,} bytes")
        print(f"offline={s['offline']}  {s['path']}")
        return
    if a.seed:
        seed(a.site, a.lat, a.lon, a.radius)
        return
    if os.getenv("DATA_OFFLINE", "1") == "1":
        p.error("Set DATA_OFFLINE=0 to fetch live, or pass --seed for placeholders")
    warm(a.site, a.lat, a.lon, a.radius)


if __name__ == "__main__":
    main()

File: lib/data/warm.ts
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { parseArgs } from "node:util";

import { diskCachePath, formEncode } from "./http";
import { NOMINATIM, OVERPASS, footprints, geocode } from "./osm";

/**
 * Warm the public-data cache.
 *
 *     DATA_OFFLINE=0 DATA_CONTACT="you@example.com" \
 *       npm run warm -- --site "Willow Park, Kilimani, Nairobi" \
 *                       --lat -1.29210 --lon 36.78270
 *
 * Fetches from Nominatim and Overpass, writes to fixtures/cache (or the KV
 * store when configured), and from then on the app and the tests run
 * offline against exactly what you fetched. Commit the cache so a demo
 * never depends on someone else's uptime.
 *
 *     npm run warm -- --seed    write placeholder fixtures so the offline
 *                               demo runs before you have fetched anything
 */

interface Target {
  site: string;
  lat: number;
  lon: number;
  radius: number;
}

async function warm({ site, lat, lon, radius }: Target): Promise<void> {
  console.log(`geocoding  ${site}`);
  const g = await geocode(site);
  console.log(
    `  ${g.confidence.padEnd(12)} ${g.latitude.toFixed(5)}, ${g.longitude.toFixed(5)}  ${g.displayName.slice(0, 70)}`,
  );

  console.log(`footprints within ${radius}m of ${lat}, ${lon}`);
  const f = await footprints(lat, lon, radius);
  console.log(
    `  ${f.buildingCount} buildings, ${f.buildingsUnderConstruction} under construction`,
  );
  if (f.landuse.length) console.log(`  land use: ${f.landuse.join(", ")}`);
}

/**
 * Placeholder responses shaped exactly like the real APIs, so the offline
 * demo runs end to end before anyone has network access. Overwritten the
 * moment you warm the cache for real.
 */
function seed({ site, lat, lon, radius }: Target): void {
  const geoUrl =
    NOMINATIM +
    "?" +
    formEncode({
      q: site,
      countrycodes: "ke",
      format: "jsonv2",
      limit: 1,
      addressdetails: 1,
    });
  const geoBody = JSON.stringify([
    {
      place_id: 297451102,
      osm_type: "way",
      osm_id: 1043318877,
      lat: (lat + 0.0011).toFixed(7),
      lon: (lon - 0.0008).toFixed(7),
      display_name: `${site}, Nairobi, Kenya`,
      category: "landuse",
      type: "residential",
      importance: 0.27,
      address: { suburb: "Kilimani", city: "Nairobi", country_code: "ke" },
    },
  ]);

  const query = `[out:json][timeout:40];
(
  way["building"](around:${radius},${lat},${lon});
  relation["building"](around:${radius},${lat},${lon});
  way["landuse"](around:${radius},${lat},${lon});
);
out center tags;`;
  const overBody = JSON.stringify({
    version: 0.6,
    generator: "Overpass API",
    elements: [
      {
        type: "way",
        id: 411229301,
        center: { lat: lat + 0.00042, lon: lon + 0.00031 },
        tags: { building: "apartments", "building:levels": "6" },
      },
      {
        type: "way",
        id: 411229774,
        center: { lat: lat - 0.00061, lon: lon + 0.0009 },
        tags: { building: "residential" },
      },
      {
        type: "way",
        id: 902114553,
        center: { lat: lat + 0.00105, lon: lon - 0.00044 },
        tags: { building: "construction", construction: "apartments" },
      },
      {
        type: "way",
        id: 118844021,
        tags: { landuse: "residential", name: "Kilimani" },
      },
    ],
  });

  const targets: Array<[string, Buffer | null, string]> = [
    [geoUrl, null, geoBody],
    [OVERPASS, Buffer.from(formEncode({ data: query })), overBody],
  ];
  for (const [url, payload, body] of targets) {
    const path = diskCachePath(url, payload);
    mkdirSync(dirname(path), { recursive: true });
    writeFileSync(path, body);
    console.log(`seeded  ${path}`);
  }
}

async function main(): Promise<void> {
  const { values } = parseArgs({
    options: {
      site: { type: "string", default: "Willow Park Block A, Kilimani" },
      lat: { type: "string", default: "-1.29210" },
      lon: { type: "string", default: "36.78270" },
      radius: { type: "string", default: "200" },
      seed: { type: "boolean", default: false },
    },
  });
  const target: Target = {
    site: values.site!,
    lat: Number.parseFloat(values.lat!),
    lon: Number.parseFloat(values.lon!),
    radius: Number.parseInt(values.radius!, 10),
  };

  if (values.seed) {
    seed(target);
    return;
  }
  if ((process.env.DATA_OFFLINE ?? "1") === "1") {
    console.error("Set DATA_OFFLINE=0 to fetch live, or pass --seed for placeholders");
    process.exitCode = 1;
    return;
  }
  await warm(target);
}

main().catch((error: unknown) => {
  console.error(error instanceof Error ? error.message : error);
  process.exitCode = 1;
});
